### PR TITLE
De-flake `UserRestartTest`

### PR DESCRIPTION
**Pull request by basil**
### Problem

`hudson.model.UserRestartTest#badSerialization` flaked with:

```
java.lang.AssertionError: expected:<pqhacker> but was:<null>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:120)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at hudson.model.UserRestartTest.lambda$badSerialization$3(UserRestartTest.java:84)
	at org.jvnet.hudson.test.JenkinsSessionRule$2.evaluate(JenkinsSessionRule.java:104)
	at org.jvnet.hudson.test.JenkinsRule$1.evaluate(JenkinsRule.java:618)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:299)
	at org.junit.internal.runners.statements.FailOnTimeout$CallableStatement.call(FailOnTimeout.java:293)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

### Evaluation

The cause is similar to #7199, but in this case rather than a race between two jobs being loaded it is a race between a job being loaded and the user ID mapper being initialized. I could reliably reproduce this race by adding:

```java
diff --git a/core/src/main/java/hudson/model/UserIdMapper.java b/core/src/main/java/hudson/model/UserIdMapper.java
index 93e6a03fd0..cb12b50dff 100644
--- a/core/src/main/java/hudson/model/UserIdMapper.java
+++ b/core/src/main/java/hudson/model/UserIdMapper.java
@@ -71,6 +71,11 @@ public class UserIdMapper {
 
     @Initializer(after = InitMilestone.PLUGINS_STARTED, before = InitMilestone.JOB_LOADED)
     public File init() throws IOException {
+        try {
+            Thread.sleep(10000);
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
         usersDirectory = createUsersDirectoryAsNeeded();
         load();
         return usersDirectory;
```

This delays the initialization of the user ID mapper, allowing job loading to proceed without a fully initialized user ID mapper and exposing us to the race. Job loading calls `hudson.model.User.Replacer#readResolve`, which implicitly depends on the user ID having been mapped, which is not guaranteed because it is happening concurrently, and when the sleep is added we lose the race and get null, then the test fails.

### Solution

I solved this the same way as in #7199, not by changing the production code but by making the test more robust. Making the production code more robust is risky, and the production code path only exists as a best-effort workaround for a problem that was actually created much earlier when the object was serialized, anyway.

### Testing done

With the sleep in place, I could reliably reproduce the failure before this PR and could no longer reproduce it after this PR.

### Proposed changelog entries

N/A

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7449"></a>

### CODE REVIEW

File: core/src/main/java/hudson/model/User.java
Comment (basil) [inline core/src/main/java/hudson/model/User.java:1069]: Consistent with the comment in https://github.com/jenkinsci/jenkins/blob/86786b035c226276792068c66d0a8ae7468d33b6/core/src/main/java/hudson/model/AbstractItem.java#L643 (which I did _not_ write).

File: test/src/test/java/hudson/model/UserRestartTest.java
Comment (basil) [inline test/src/test/java/hudson/model/UserRestartTest.java:90]: Consistent with what was done in #7199.

**Review by MarkEWaite — APPROVED** [review]

**Comment by basil** [comment]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by NotMyFault — APPROVED** [review]

**Pull request merged**
